zebra-network/src/peer_set/initialize.rs

```diff
@@ -904,20 +924,48 @@ where
     }
 }
 
-/// Mark `addr` as a failed peer.
-async fn report_failed(address_book: Arc<std::sync::Mutex<AddressBook>>, addr: MetaAddr) {
-    let addr = MetaAddr::new_errored(addr.addr, addr.services);
+/// Try to get more peers using `candidates`, then queue a connection attempt using `demand_tx`.
+/// If there were no new peers, the connection attempt is skipped.
+#[instrument(skip(candidates, demand_tx))]
+async fn crawl<S>(
```

Comment:
That seems like it would be harder to read, because the code would be indented, and the closures or functions could look like they execute where they're defined.